core/server/src/main/java/alluxio/worker/file/FileDataManager.java

```diff
@@ -292,34 +298,57 @@ public void persistFile(long fileId, List<Long> blockIds) throws IOException {
     AlluxioURI uri = new AlluxioURI(fileInfo.getPath());
-    String dstPath = PathUtils.concatPath(ufsRoot, fileInfo.getPath());
+    AlluxioURI dstPath = new AlluxioURI(PathUtils.concatPath(ufsRoot, fileInfo.getPath()));
     LOG.info("persist file {} at {}", fileId, dstPath);
     String parentPath = PathUtils.concatPath(ufsRoot, uri.getParent().getPath());
     // creates the parent folder if it does not exist
     if (!mUfs.isDirectory(parentPath)) {
-      final int maxRetry = 10;
-      int numRetry = 0;
-      // TODO(peis): Retry only if we are making progress.
-      // TODO(chaomin): figure out a way to get parent permission in Alluxio namespace.
-      for (; numRetry < maxRetry; numRetry++) {
-        if (mUfs.mkdirs(parentPath) || mUfs.isDirectory(parentPath)) {
-          break;
+      FileSystem fs = FileSystem.Factory.get();
+      // Create ancestor directories from top to the bottom. We cannot use recursive create parents
+      // here because the permission for the ancestors can be different.
+      Stack<Pair<String, MkdirsOptions>> ufsDirsToMakeWithOptions = new Stack<>();
+      AlluxioURI curAlluxioPath = uri.getParent();
+      AlluxioURI curUfsPath = dstPath.getParent();
+      // Stop at the Alluxio root because the mapped directory of Alluxio root in UFS may not exist.
+      while (!mUfs.isDirectory(curUfsPath.toString()) && curAlluxioPath != null) {
+        URIStatus curDirStatus;
+        curDirStatus = fs.getStatus(curAlluxioPath);
+        Permission perm = new Permission(curDirStatus.getOwner(), curDirStatus.getGroup(),
+            (short) curDirStatus.getMode());
+        ufsDirsToMakeWithOptions.push(new Pair<>(curUfsPath.toString(),
+            MkdirsOptions.defaults().setCreateParent(false).setPermission(perm)));
+        curAlluxioPath = curAlluxioPath.getParent();
+        curUfsPath = curUfsPath.getParent();
+      }
+      while (!ufsDirsToMakeWithOptions.empty()) {
+        Pair<String, MkdirsOptions> ufsDirAndPerm = ufsDirsToMakeWithOptions.pop();
+        if (!createUfsDirWithRetry(ufsDirAndPerm.getFirst(), ufsDirAndPerm.getSecond())) {
+          throw new IOException("Failed to create dir: " + ufsDirAndPerm.getFirst());
         }
-        // The parentPath can be created between the exists check and mkdirs call by other threads.
-        LOG.warn("Failed to create dir: {}, retrying", parentPath);
       }
-      if (numRetry == maxRetry && !mUfs.isDirectory(parentPath)) {
-        throw new IOException(
-            String.format("Failed to create dir: %s after %d retries.", parentPath, numRetry));
+    }
+    return dstPath.toString();
+  }
+
+  private boolean createUfsDirWithRetry(String dirPath, MkdirsOptions options) throws IOException {
+    final int maxRetry = 10;
+    int numRetry = 0;
+    // TODO(peis): Retry only if we are making progress.
+    for (; numRetry < maxRetry; numRetry++) {
+      if (mUfs.isDirectory(dirPath) || mUfs.mkdirs(dirPath, options)) {
```

Comment:
We had to retry 10 times because async-persist needed to create dir recursively and the recursive dir creation is not atomic. Do you still this to retry 10 times after this change?